### ISSUE TITLE
Distinct samples of training data

### ISSUE THREAD

**Issue by leowe** [google/ffn#16]
Is it possible to train the ffn with several distinct pieces of data with each having its own bounding box?

**Comment by subeeshvasu** [comment]
I am also trying to do the same. After inspecting the code, I figured out the following solution, to train with more than one data volumes.
Let's assume that we have two training volumes at hand.
- First, generate the coordinates for both volumes, but with different tags. i.e., while using 'build_coordinates.py' we could use "validation1" as the tag for volume 1 and "validation2" as the tag for volume 2.
- The generated coordinates should be named as "tf_record_file-00000-of-00002" and tf_record_file-00001-of-00002, since the code is designed to read files in this format when there are more than one coordinate file. 
- Change the argument: 
**train_coords** 
from 
"tf_record_file" 
to 
"tf_record_file@2", 
where 2 is the number of coordinate files
- Next, change the arguments corresponding to input volume and corresponding labels: 
**data_volumes** 
from 
validation1:address_to_volume_1.h5:raw
to
validation1:address_to_volume_1.h5:raw,validation2:address_to_volume_2.h5:raw

Repeat the above step for 
**label_volumes**
too

- Now run the training code as usual. This was expected to work. However, when I tried, for some unknown reason, the training code is not alternating between the two train volumes. Because of random shuffling, each time when I run, i can see that the training code is able to access both of the volumes. But, once it starts training, it doesn't alternate between the two volumes. I am yet to figure out the reason for this. If anyone has an answer to this problem, please let me know.

Just to be clear, I will repeat the issue I am facing:
If I run the code with the above modifications,
run 1 -> will use only validation 1  (or 2 depending on the output of random shuffle) volume
run 2 -> will use only validation 2 (or 1 depending on the output of random shuffle) volume

 ffn/training/inputs - the line **tf.TFRecordReader(options=record_options).read(filename_queue)** - is expected to alternate between the names of both volumes (since filename_queue is generates using tf.train.string_input_producer). But this is not happening when I run. In my run this line consistently output the same volume name!!!

Can anyone help me to figure out a solution?

**Comment by mk123qwe** [comment]
> I am also trying to do the same. After inspecting the code, I figured out the following solution, to train with more than one data volumes.
> Let's assume that we have two training volumes at hand.
> 
> * First, generate the coordinates for both volumes, but with different tags. i.e., while using 'build_coordinates.py' we could use "validation1" as the tag for volume 1 and "validation2" as the tag for volume 2.
>   eg. python build_coordinates.py --partition_volumes validation1:address_to_volume1_af.h5:af_file,validation2:address_to_volume2_af.h5:af_file --coordinate_output ..
> * Next, change the arguments corresponding to input volume and corresponding labels:
>   **data_volumes**
>   from
>   validation1:address_to_volume_1.h5:raw
>   to
>   validation1:address_to_volume_1.h5:raw,validation2:address_to_volume_2.h5:raw
> 
> Repeat the above step for
> **label_volumes**
> too
> 
> * Now run the training code as usual.
How to save the model according to the set conditions, not periodically.
I noticed that MonitoredSession is a very inefficient method.

**Comment by mk123qwe** [comment]
> I am also trying to do the same. After inspecting the code, I figured out the following solution, to train with more than one data volumes.
> Let's assume that we have two training volumes at hand.
> 
> * First, generate the coordinates for both volumes, but with different tags. i.e., while using 'build_coordinates.py' we could use "validation1" as the tag for volume 1 and "validation2" as the tag for volume 2.
>   eg. python build_coordinates.py --partition_volumes validation1:address_to_volume1_af.h5:af_file,validation2:address_to_volume2_af.h5:af_file --coordinate_output ..
> * Next, change the arguments corresponding to input volume and corresponding labels:
>   **data_volumes**
>   from
>   validation1:address_to_volume_1.h5:raw
>   to
>   validation1:address_to_volume_1.h5:raw,validation2:address_to_volume_2.h5:raw
> 
> Repeat the above step for
> **label_volumes**
> too
> 
> * Now run the training code as usual.

Hello, I'm confused that this program has no verification set.

**Comment by subeeshvasu** [comment]
> > I am also trying to do the same. After inspecting the code, I figured out the following solution, to train with more than one data volumes.
> > Let's assume that we have two training volumes at hand.
> > 
> > * First, generate the coordinates for both volumes, but with different tags. i.e., while using 'build_coordinates.py' we could use "validation1" as the tag for volume 1 and "validation2" as the tag for volume 2.
> >   eg. python build_coordinates.py --partition_volumes validation1:address_to_volume1_af.h5:af_file,validation2:address_to_volume2_af.h5:af_file --coordinate_output ..
> > * Next, change the arguments corresponding to input volume and corresponding labels:
> >   **data_volumes**
> >   from
> >   validation1:address_to_volume_1.h5:raw
> >   to
> >   validation1:address_to_volume_1.h5:raw,validation2:address_to_volume_2.h5:raw
> > 
> > Repeat the above step for
> > **label_volumes**
> > too
> > 
> > * Now run the training code as usual.
> 
> Hello, I'm confused that this program has no verification set.
I didn't understand. You meant: "No setup/data to check monitor the improvement in segmentation?". If yes, I agree. I would also like to figure out a way to save the network weights based on some sort of quantitative evaluation. I haven't explored that direction yet. Let me know if you figure out a way to do it. Thanks.

**Comment by mk123qwe** [comment]
> > > I am also trying to do the same. After inspecting the code, I figured out the following solution, to train with more than one data volumes.
> > > Let's assume that we have two training volumes at hand.
> > > 
> > > * First, generate the coordinates for both volumes, but with different tags. i.e., while using 'build_coordinates.py' we could use "validation1" as the tag for volume 1 and "validation2" as the tag for volume 2.
> > >   eg. python build_coordinates.py --partition_volumes validation1:address_to_volume1_af.h5:af_file,validation2:address_to_volume2_af.h5:af_file --coordinate_output ..
> > > * Next, change the arguments corresponding to input volume and corresponding labels:
> > >   **data_volumes**
> > >   from
> > >   validation1:address_to_volume_1.h5:raw
> > >   to
> > >   validation1:address_to_volume_1.h5:raw,validation2:address_to_volume_2.h5:raw
> > > 
> > > Repeat the above step for
> > > **label_volumes**
> > > too
> > > 
> > > * Now run the training code as usual.
> > 
> > 
> > Hello, I'm confused that this program has no verification set.
> > I didn't understand. You meant: "No setup/data to check monitor the improvement in segmentation?". If yes, I agree. I would also like to figure out a way to save the network weights based on some sort of quantitative evaluation. I haven't explored that direction yet. Let me know if you figure out a way to do it. Thanks.

I don't understand enough about FFN training process, which is different from what I know about segmentation tasks.
"MonitoredTrainingSession" is different from Session.You can try "Supervisor".
